barretenberg/cpp/src/barretenberg/honk/instance/instance.hpp

```diff
@@ -0,0 +1,114 @@
+#include "barretenberg/honk/flavor/ultra.hpp"
+#include "barretenberg/honk/flavor/ultra_grumpkin.hpp"
+#include "barretenberg/honk/proof_system/folding_result.hpp"
+#include "barretenberg/proof_system/composer/composer_lib.hpp"
+#include "barretenberg/proof_system/flavor/flavor.hpp"
+#include "barretenberg/proof_system/relations/relation_parameters.hpp"
+#include "barretenberg/srs/factories/file_crs_factory.hpp"
+#include <cstddef>
+#include <memory>
+#include <utility>
+#include <vector>
+namespace proof_system::honk {
+/**
+ * @brief  An Instance is normally constructed from a finalised circuit and it's role is to compute all the polynomials
+ * involved in creating a proof and, if requested, the verification key.
+ * In case of folded Instance, this will be created from the FoldingResult, the aggregated work from the folding prover
+ * and verifier. More specifically, a folded instance will be constructed from the complete set of folded polynomials
+ * and folded public inputs and its FoldingParams are expected to be non-zero
+ *
+ */
+
+template <UltraFlavor Flavor> class Instance_ {
+  public:
+    using Circuit = typename Flavor::CircuitBuilder;
+    using ProvingKey = typename Flavor::ProvingKey;
+    using VerificationKey = typename Flavor::VerificationKey;
+    using CommitmentKey = typename Flavor::CommitmentKey;
+    using FF = typename Flavor::FF;
+    using FoldingParameters = typename Flavor::FoldingParameters;
+    using ProverPolynomials = typename Flavor::ProverPolynomials;
+    using Polynomial = typename Flavor::Polynomial;
+
+    // offset due to placing zero wires at the start of execution trace
+    static constexpr size_t num_zero_rows = Flavor::has_zero_row ? 1 : 0;
+
+    static constexpr std::string_view NAME_STRING = "UltraHonk";
+    static constexpr size_t NUM_WIRES = Circuit::NUM_WIRES;
+    std::shared_ptr<ProvingKey> proving_key;
+    std::shared_ptr<VerificationKey> verification_key;
+    std::shared_ptr<CommitmentKey> commitment_key;
+
+    ProverPolynomials prover_polynomials;
+
+    // The number of public inputs has to be the same for all instances because they are
+    // folded element by element.
+    std::vector<FF> public_inputs;
+    size_t pub_inputs_offset;
+
+    proof_system::RelationParameters<FF> relation_parameters;
+
+    std::vector<uint32_t> recursive_proof_public_input_indices;
+    bool contains_recursive_proof = false;
+    bool computed_witness = false;
+    size_t total_num_gates = 0; // num_gates + num_pub_inputs + tables + zero_row_offset (used to compute dyadic size)
+    size_t dyadic_circuit_size = 0; // final power-of-2 circuit size
+    size_t lookups_size = 0;        // total number of lookup gates
+    size_t tables_size = 0;         // total number of table entries
+    size_t num_public_inputs = 0;
+    size_t num_ecc_op_gates = 0;
+
+    FoldingParameters folding_params;
+    // Used by the prover for domain separation in the transcript
+    uint32_t index;
```

Comment:
This `index` usage feels kinda bad and brittle. When you create the vector of shared pointers to instances, why can't we just use the index of the pointer in that vector?

Reply:
I've mocked out related things in my combiner work, and I ended up defining an `Instances` object that wraps multiple instances. What do you think about implementing that and having that class manage the domain separation? https://github.com/AztecProtocol/aztec-packages/blob/e74d800bc814cc830d419db0415c837aacb41fe2/circuits/cpp/barretenberg/cpp/src/barretenberg/protogalaxy/combiner.hpp#L11

Reply:
Sounds good conceptually but I'd add this as TODO work and do it in a subsequent PR, maybe we get the combiner merged and then I work on smoothing things out. Wdyt?

Reply:
What about having a lightweight `Instances` class that encapsulates `NUM_`-many` Instance` objects. I've tentatively added a TODO (+issue: https://github.com/AztecProtocol/barretenberg/issues/725) to smoothen things out?